torch/csrc/autograd/python_variable.cpp

```diff
@@ -236,7 +236,107 @@ PyObject * THPVariable_Wrap(at::TensorBase var)
       (PyTypeObject*)THPVariableClass, std::move(var), status);
 }
 
+bool isResurrectable(THPVariable* self) {
+  // We want to divide this check into 2 cases.
+
+  // 1. C++ owns PyObject (in this case, self->cdata.unsafeIsBorrowed() is
+  // true). You might think that in this case, it is impossible for tp_clear to
+  // be called: surely the C++ reference to the PyObject is keeping it live? And
+  // you'd be right! In fact, when C++ owns the PyObject, we have an invariant
+  // that the refcount on the PyObject should be precisely one (because if you
+  // take out another reference to the PyObject, we're supposed to flip the
+  // ownership pointer back). In reality, you can violate this invariant
+  // temporarily with weak references, so we don't test for it in asserts.
+
+  // 2. PyObject owns C++ (in this case, self->cdata.unsafeIsBorrowed() is
+  // false). In this case, tp_clear can get called if the PyObject is referenced
+  // from a dead cycle, and nowhere else. But if resurrection did not occur,
+  // then the reference to C++ from the PyObject must be the ONLY reference to
+  // the C++ object.
+  if (self->cdata.unsafeIsBorrowed()) {
```

Comment:
Why isn't this case an assert? We should never check if the PyObject is ressurectable when the C++ owns it right?

Reply:
Hmm, we use this function in a couple different places, including in tp_clear (where it’s possible that C++ *does* own the pyobject).